### PR TITLE
stats: remove invalid range when split

### PR DESCRIPTION
**Pull request by lamxTyler**
PTAL @coocood @winoros

### CODE REVIEW

File: statistics/histogram.go [hunk 1]
```diff
@@ -441,6 +457,10 @@ func (hg *Histogram) SplitRange(ranges []*ranger.NewRange) []*ranger.NewRange {
 			ranges[0].LowExclude = true
 		}
 	}
+	valid := validRange(split[len(split)-1])
```
Comment (coocood) [inline statistics/histogram.go]: can other range rather than the last one be invalid?

Comment (lamxTyler) [reply statistics/histogram.go]: First, all of the `ranges` are valid, then for the split range, we can make sure that they are valid, so the only the last one could be invalid.

File: statistics/histogram.go [hunk 2]
```diff
@@ -392,10 +392,29 @@ func (hg *Histogram) getIncreaseFactor(totalCount int64) float64 {
 	return float64(totalCount) / float64(columnCount)
 }
 
+func validRange(ran *ranger.NewRange) bool {
+	var low, high []byte
+	if ran.LowVal[0].Kind() == types.KindBytes {
+		low, high = ran.LowVal[0].GetBytes(), ran.HighVal[0].GetBytes()
+	} else {
+		low, high = codec.EncodeInt(nil, ran.LowVal[0].GetInt64()), codec.EncodeInt(nil, ran.HighVal[0].GetInt64())
```
Comment (coocood) [inline statistics/histogram.go]: Can we just compare the int value instead of encode it?

File: statistics/histogram.go [hunk 3]
```diff
@@ -439,6 +455,9 @@ func (hg *Histogram) SplitRange(ranges []*ranger.NewRange) []*ranger.NewRange {
 				HighExclude: false})
 			ranges[0].LowVal[0] = upper
 			ranges[0].LowExclude = true
```
Comment (coocood) [inline statistics/histogram.go]: Can we enforce LowExclude to false?
By PrefixNext Bytes or increase interger?

File: statistics/histogram.go (hunk 2 above)
Comment (lamxTyler) [reply statistics/histogram.go]: Then we need to consider carefully the four combinations of low exclude and high exclude?

File: statistics/histogram.go (hunk 3 above)
Comment (lamxTyler) [reply statistics/histogram.go]: Could it possible that increase integer cause overflow?

File: statistics/histogram.go [hunk 4]
```diff
@@ -392,6 +392,22 @@ func (hg *Histogram) getIncreaseFactor(totalCount int64) float64 {
 	return float64(totalCount) / float64(columnCount)
 }
 
+func validRange(ran *ranger.NewRange) bool {
```
Comment (shenli) [inline statistics/histogram.go]: Add comment for the function.

File: statistics/update_test.go [hunk 5]
```diff
@@ -367,6 +367,11 @@ func (s *testStatsUpdateSuite) TestSplitRange(c *C) {
 			exclude: []bool{false, false, true, true},
 			result:  "[8,8],(8,10],(20,30)",
 		},
+		{
+			points:  []int64{8, 9},
```
Comment (shenli) [inline statistics/update_test.go]: Add comment for the purpose of this test case.

File: statistics/histogram.go [hunk 6]
```diff
@@ -392,6 +392,23 @@ func (hg *Histogram) getIncreaseFactor(totalCount int64) float64 {
 	return float64(totalCount) / float64(columnCount)
 }
 
+// validRange checks if the range is valid, it is used by `SplitRange` to remove the invalid range.
+func validRange(ran *ranger.NewRange) bool {
+	var low, high []byte
+	if ran.LowVal[0].Kind() == types.KindBytes {
+		low, high = ran.LowVal[0].GetBytes(), ran.HighVal[0].GetBytes()
+	} else {
+		low, high = codec.EncodeInt(nil, ran.LowVal[0].GetInt64()), codec.EncodeInt(nil, ran.HighVal[0].GetInt64())
```
Comment (shenli) [inline statistics/histogram.go]: How about uint?

Comment (lamxTyler) [reply statistics/histogram.go]: When [encoding](https://github.com/pingcap/tidb/blob/master/distsql/request_builder.go#L176) the handle key, we also only consider int.

**Comment by coocood** [comment]
LGTM

File: statistics/histogram.go (hunk 6 above)
Comment (shenli) [reply statistics/histogram.go]: Please comment it.

**Comment by shenli** [comment]
LGTM

File: statistics/histogram.go (hunk 6 above)
Comment (lamxTyler) [reply statistics/histogram.go]: Done.

Comment (winoros) [reply statistics/histogram.go]: The range for request will only contain the [MinInt64, MaxInt64] part or the [MaxInt64+1, MaxUint64] part. It's already split beyond there.

**Pull request merged**